### PR TITLE
internal: Allow building on 32bit (fixes #884)

### LINKED ISSUE #884: Not compatible with 32 bits arches
Hi,

While building on 32 bits arches, with Go 1.15, building fails with:

```
# github.com/cockroachdb/pebble
_build/src/github.com/cockroachdb/pebble/batch.go:801:13: constant 4294967296 overflows int
_build/src/github.com/cockroachdb/pebble/options.go:935:20: constant 4294967296 overflows int
```

Is it not supposed to work on armv7 and i686?

### PR DESCRIPTION
**Pull request by imsodin**
I was looking into pebble as a replacement for goleveldb in Syncthing, when our CI came up red for the 32bit cross-builds (#884). This PR fixes those build issues. However while the changes in batchskl look sane to me, I am obviously not familiar with the code, and for manual/rawalloc I have no clue about how the affected methods are used at all, i.e. if the limit for the array size that's < max_uint32 is likely to be problematic or not.  
Basically my question now is for people knowing the code-base: Any gotchas with these changes? And more generally, given 32bit has been clearly broken since a while, what's your take on the general stability/usability of pebble on 32bit?

### CODE REVIEW

**Comment by holiman** [comment]
We are also looking into pebble as a replacement for leveldb in go-ethereum, and, just like @imsodin ,the 32bit cross build issue is breaking our CI. Getting a resolution on this would be great -- a review or just answer to the questions posted already: 

>  Any gotchas with these changes? And more generally, given 32bit has been clearly broken since a while, what's your take on the general stability/usability of pebble on 32bit?

Thanks!

**Comment by imsodin** [comment]
They gave the answer in the linked ticket, unfortunately negative:

https://github.com/cockroachdb/pebble/issues/1575#issuecomment-1230872230
> We only support 64-bit for now, so I'm going to close this out.

If anyone news a maintained, go-only key-value store that supports 32bit, I'd appreciate pointers (goleveldb had a bunch of new contributions lately, maybe it's going to improve still).

**Comment by fjl** [comment]
IMHO it would be very good to get this PR in. It's only 50 LOC of change, and will at least allow us to compile pebble on 32bit.

**Review by jbowens — COMMENTED** [review]
We're not opposed to 32-bit support as long as it doesn't add any additional maintenance burden.

> Reviewed 4 of 6 files at r1, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1619#-:-NAyjIMM7SD799aEuxzU:b-fnahx3)* status: all files reviewed (commit messages unreviewed), 4 unresolved discussions (waiting on @imsodin)

___
*[`internal/batchskl/skl.go` line 91 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1619#-NAycC488plQ4zP2-OHW:-NAycC49BCU8gGMHJTcL:b6rn10t) ([raw file](https://github.com/cockroachdb/pebble/blob/47bb0dfc973aec3ffdd597c639cf050917b0668e/internal/batchskl/skl.go#L91)):*
> ```Go
> 
> func init() {
> 	// Slice size is limited by arch dependendent maximum of int and size of
> ```

nit: s/dependendent/dependent/
___
*[`internal/batchskl/skl.go` line 92 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1619#-NAyaVyu0y_MyE2d5CTo:-NAyaVyu0y_MyE2d5CTp:b5kpnyo) ([raw file](https://github.com/cockroachdb/pebble/blob/47bb0dfc973aec3ffdd597c639cf050917b0668e/internal/batchskl/skl.go#L92)):*
> ```Go
> func init() {
> 	// Slice size is limited by arch dependendent maximum of int and size of
> 	// array element (byte). Cannot use math.MaxInt as that's notassignable to
> ```

nit: s/notassignable/not assignable/
___
*[`internal/batchskl/skl.go` line 94 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1619#-NAyfdF91UfbJ-wgsJrX:-NAyfdF91UfbJ-wgsJrY:bktr5xn) ([raw file](https://github.com/cockroachdb/pebble/blob/47bb0dfc973aec3ffdd597c639cf050917b0668e/internal/batchskl/skl.go#L94)):*
> ```Go
> 	// array element (byte). Cannot use math.MaxInt as that's notassignable to
> 	// uint32 on 64bit archs.
> 	if intBits := unsafe.Sizeof(int(0)) * 8; intBits > 32 {
> ```

I think the runtime uses something like this expression, which can be lifted into a const. mind doing that here?
```
const wordSizeBits = 32 << (^uintptr(0) >> 63) // equivalent to unsafe.Sizeof(uintptr(0))
```
___
*[`internal/batchskl/skl.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1619#-NAyh3jZ5QaMWtDNpSUz:-NAyh3jZ5QaMWtDNpSV-:blov3s7) ([raw file](https://github.com/cockroachdb/pebble/blob/47bb0dfc973aec3ffdd597c639cf050917b0668e/internal/batchskl/skl.go#L97)):*
> ```Go
> 		maxNodesSize = math.MaxUint32
> 	} else {
> 		maxNodesSize = 1 << (intBits - 1) / 2
> ```

can this be `manual.MaxArrayLen`?

**Comment by fjl** [comment]
Thanks for reviewing it, @jbowens!

**Review by imsodin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1619)* status: 5 of 6 files reviewed, 4 unresolved discussions (waiting on @imsodin and @jbowens)

___
*[`internal/batchskl/skl.go` line 94 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1619#-NAyfdF91UfbJ-wgsJrX:-NB4Cwsp7KgsgindfIYg:bq3k0z1) ([raw file](https://github.com/cockroachdb/pebble/blob/391ea34a661762ada1157a9e061bc71761541876/internal/batchskl/skl.go#L94)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think the runtime uses something like this expression, which can be lifted into a const. mind doing that here?
```
const wordSizeBits = 32 << (^uintptr(0) >> 63) // equivalent to unsafe.Sizeof(uintptr(0))
```

</blockquote></details>

I'have just simplified it to one simple constant based on slice size limit and mem allocation limits (lowest for both coincide here) - added a comment.
___
*[`internal/batchskl/skl.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1619#-NAyh3jZ5QaMWtDNpSUz:-NB4ChL63HzWUkao9jnQ:b7xgkkz) ([raw file](https://github.com/cockroachdb/pebble/blob/391ea34a661762ada1157a9e061bc71761541876/internal/batchskl/skl.go#L97)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

can this be `manual.MaxArrayLen`?
</blockquote></details>

We can't use that here, as we are constrained to uint32 while for 64bit manual.MaxArrayLen is much larger than that.

**Comment by imsodin** [comment]
Addressed or answered review comments.

This needs CI with `GOARCH=i386` as well-  something else changed since I originally PRed on mainline branch, it's currently failing (didn't have time to look at it yet).

**Comment by holiman** [comment]
We're still hoping to get this PR in. We cannot really begin making use of Pebble as backend as long as it cannot be compiled on 32bit platforms, and we're looking forward to start using Pebble!

**Comment by holiman** [comment]
@imsodin Just to clarify: my comment was not directed at you. You have done good job getting it thus far, and fwict have addressed the review-comments. 

Seeing as "All checks have passed", I was under the impression that this is on re-review stage now, and that there were not much non-maintainers like you or me can do at the moment, hence the ping.

**Comment by imsodin** [comment]
My main point, not sure if it that was clear, is that with this PR it still doesn't build on 32bit.

**Comment by nicktrav** [comment]
Hello all - Apologies for the delayed response. Thank you for your patience here, and thank you @imsodin for your work on this to-date.

I’d like to provide some additional context around 32-bit support. While Pebble will (for the most part) work in standalone setups (i.e. when not part of a CockroachDB deployment), it is developed and tested (almost exclusively) with CockroachDB in mind. Given that CockroachDB currently does not target 32-bit architectures, there has historically been no need for such support in Pebble. Our roadmap is currently driven by Cockroach features. [Here is a tracking issue for 32-bit support in CockroachDB](https://github.com/cockroachdb/cockroach/issues/8179) in case anyone is interested.

I think the best way forward here would be for someone to pick this up if there is still interest, rebase on latest `master` and address the build issues mentioned. We can also look at amending the commit to build / test a 32-bit version of the binary. Once everything is passing, we can take another look at the patch.

Thanks again for your patience.

**Comment by holiman** [comment]
We're working on it,.cc @jwasinger
